pkg/storageincentives/nodestatus.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package storageincentives
+
+import (
+	"context"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethersphere/bee/pkg/log"
+	"github.com/ethersphere/bee/pkg/settlement/swap/erc20"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"math/big"
+	"sync"
+)
+
+const redistributionStatusKey = "redistribution_state_"
+const loggerNameNode = "nodestatus"
+
+type NodeState struct {
+	stateStore     storage.StateStorer
+	erc20Service   erc20.Service
+	logger         log.Logger
+	overlay        swarm.Address
+	mtx            sync.Mutex
+	nodeStatus     NodeStatus
+	initialBalance *big.Int
+}
+
+// NodeStatus provide internal status of the nodes in the redistribution game
+type NodeStatus struct {
+	State  PhaseType `json:"state"`
+	Round  uint64    `json:"round"`
+	Block  uint64    `json:"block"`
+	Reward *big.Int  `json:"reward"`
+	Fees   *big.Int  `json:"fees"`
+}
+
+func NewNode(logger log.Logger, stateStore storage.StateStorer, erc20Service erc20.Service) NodeState {
+	return NodeState{
+		stateStore:     stateStore,
+		erc20Service:   erc20Service,
+		logger:         logger.WithName(loggerNameNode).Register(),
+		initialBalance: big.NewInt(0),
+		nodeStatus: NodeStatus{
+			Round:  0,
+			Block:  0,
+			Reward: nil,
+			Fees:   nil,
+		},
+	}
+}
+
+func (n *NodeState) SetPhase(p PhaseType) {
```

Comment:
it is problematic to include the "frozen" state in the function, because when the next phase happens, for example, if the phase is currently "frozen" and then we move on to the "claim" phase, it will override the "frozen" state.

this why SetPhase should only include the claim, commit, and reveal phases, and you should have a new function and field field for freeze and winner.

Reply:
My understanding is that if the node is `Frozen`, it cannot participate in the game. Separating Phase and State would be confusing.

Reply:
it cannot participate, yes, but the agent will keep running, going through each phase, and updating this field

Reply:
Okay, so I have separated them now.